### PR TITLE
[fix] Improve handling of illegal vFAT filenames.

### PR DESCRIPTION
**Pull request by michal-mnich**
This change fixes an issue where a file/directory is not created when its name contains trailing spaces or dots. The issue has been observed on a Kobo device, but probably also applies to other systems using vFAT.

We also might want to consider handling other illegal vFAT filenames like `NUL` or `AUX`.

Please see the [official Microsoft documentation](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions) for more details on the naming conventions of vFAT (this change addresses the last bullet point).

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/14043)

### CODE REVIEW

File: frontend/util.lua [hunk 1]
```diff
@@ -948,14 +948,17 @@ end
...
 -- These characters are problematic on Windows filesystems. On Linux only
 -- <code>/</code> poses a problem.
 ---- @string str filename
 ---- @treturn string sanitized filename
 local function replaceAllInvalidChars(str)
     if str then
-        return str:gsub('[\\/:*?"<>|]', '_')
+        str = str:gsub('[\\/:*?"<>|]', '_')
+        str = str:gsub("[%.%s]+$", "")
+        return str
```
Comment (Frenzie) [inline frontend/util.lua]: It might be more efficient (in a way that doesn't matter :-)) to write it something like this. But is it clearer. :thinking: 
```suggestion
        return str:gsub('[\\/:*?"<>|]', '_')
                  :gsub("[%.%s]+$", "")
```

**Comment by Frenzie** [comment]
> This change fixes an issue where a file/directory is not created when its name contains trailing spaces or dots.

It would seem that trailing dots are created, but the dots generally aren't shown and the name counts as a duplicate to without trailing dots or a different number of trailing dots. That's not an objection to stripping them or anything, mind.

**Comment by michal-mnich** [comment]
I see your point. Actually, the current implementation would also treat `file?` and `file:` as duplicates (since both of them will become `file_`), similarly for other pairs of illegal characters. I think that a complete resolution of this is impossible, since the user could theoretically name the file to any of our mitigation attempts :). But maybe it would help to replace the illegal characters with their encoding of some kind (URL-like, UTF8, maybe even ASCII)? It would make an accidental collision a bit less likely.

**Comment by Frenzie** [comment]
No, I don't consider that a concern. I was describing the behavior of the /mnt/onboard filesystem on Kobo for the sake of accuracy. Only trailing spaces are problematic, give or take the edge case of trailing spaces followed by trailing dots. Creating a folder with the name `test.` works with any number of `.` (as long as you stay under 255 characters total I assume). But `test ` or `test .` will not.

File: frontend/util.lua [hunk 2]
```diff
@@ -948,14 +948,17 @@ end
...
+-- and removes trailing dots and spaces, in line with <https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions>.
 -- These characters are problematic on Windows filesystems. On Linux only
 -- <code>/</code> poses a problem.
 ---- @string str filename
 ---- @treturn string sanitized filename
 local function replaceAllInvalidChars(str)
     if str then
-        return str:gsub('[\\/:*?"<>|]', '_')
+        str = str:gsub('[\\/:*?"<>|]', '_')
+        str = str:gsub("[%.%s]+$", "")
```
Comment (Frenzie) [inline frontend/util.lua]: In any event, the escape isn't necessary here.
```suggestion
        str = str:gsub("[.%s]+$", "")
```

Comment (michal-mnich) [reply frontend/util.lua]: Why so? In the [docs](https://www.lua.org/pil/20.2.html) it says:

> Some characters, called magic characters, have special meanings when used in a pattern. The magic characters are
( ) . % + - * ? [ ^ $
The character '%' works as an escape for those magic characters. So, '%.' matches a dot; '%%' matches the character '%' itself.

Won't the dot match every character and make bad things happen :)?

Comment (Frenzie) [reply frontend/util.lua]: Inside [], most of those are just characters. Except of course for %, ^ and -.

Comment (Frenzie) [reply frontend/util.lua]: Or put another way, in `abcd.` the . means anything. But `abcd[.]` is equivalent to `abcd%.`.
